torch/nn/modules/transformer.py

```diff
@@ -196,16 +197,17 @@ class TransformerDecoder(Module):
         >>> tgt = torch.rand(20, 32, 512)
         >>> out = transformer_decoder(tgt, memory)
     """
+    __constants__ = ['norm']
 
     def __init__(self, decoder_layer, num_layers, norm=None):
         super(TransformerDecoder, self).__init__()
         self.layers = _get_clones(decoder_layer, num_layers)
-        self.num_layers = num_layers
```

Comment:
This is a BC breaking

Reply:
This seems like redundant info since someone can just do `len(self.layers)` to get the same value. Since this change is BC-breaking anyways I think this cleanup is warranted

Reply:
Since I'm targeting to cover BC, please add those two attributes back and I could approve it soon.

Reply:
`self.num_layers` is added to Encoder but not Decoder.